scripts/extensions/videoEditor/src/VideoTimeline/VideoTimeline.tsx

```diff
@@ -0,0 +1,296 @@
+        }
+
+        if (currentTime <= this.state.trim.end) {
+            this.setState({currentTime: currentTime});
+        } else if (this.state.trim.end > 0) {
+            this.setState({currentTime: this.state.trim.end});
+            this.props.video.pause();
+        }
+    }
+
+    // drag and drop left and right bar.
+    handleDragStart(e: React.DragEvent<HTMLDivElement>) {
+        // hide drag ghost image
+        const img = new Image();
+
+        img.src = 'data:image/gif;base64,R0lGODlhAQABAIAAAAUEBAAAACwAAAAAAQABAAACAkQBADs=';
+        e.dataTransfer.setDragImage(img, 0, 0);
+        e.dataTransfer.setData('text/plain', '');
+    }
+
+    handleDragEnd() {
+        this.setVideoCurrentTime(this.positionX);
+        this.props.onTrim(round(this.state.trim.start), round(this.state.trim.end));
+    }
+
+    // returns a number between 0 and 1
+    // e.g. if a video is 60 seconds position 0.25 would represent a position at 15 seconds
+    getPositionInBar(pX: number): number {
+        const controlbar = this.controlbar.current;
+
+        if (controlbar == null) {
+            throw new Error('control bar element is not present');
+        }
+
+        var position = (pX - controlbar.getBoundingClientRect().left) / controlbar.offsetWidth;
+
+        if (position > 1) {
+            return 1;
+        } else if (position < 0) {
+            return 0;
+        } else {
+            return Math.floor(position * 1000) / 1000;
+        }
+    }
+
+    handleTimelineClick(e: React.MouseEvent<HTMLDivElement>) {
+        let time = this.setVideoCurrentTime(e.clientX);
+
+        if (time < this.state.trim.start) {
+            this.props.onTrim(round(time), round(this.state.trim.end));
+            this.updateTrim(time, this.state.trim.end);
+        }
+        if (time > this.state.trim.end) {
+            this.props.onTrim(round(this.state.trim.start), round(time));
+            this.updateTrim(this.state.trim.start, time);
+        }
+    }
+
+    setVideoCurrentTime(pX: number) {
+        let time = this.getPositionInBar(pX) * this.props.video.duration;
+
+        this.props.video.currentTime = time;
+        this.setState({currentTime: time});
+        return time;
+    }
+
+    render() {
+        const {getClass} = this.props;
+        const video = this.props.video;
+        const left = video ? `${(this.state.trim.start / video.duration) * 100}%` : '0%';
+        const right = video ? `${(1 - this.state.trim.end / video.duration) * 100}%` : '0%';
+        const thumbnailWidth = Math.round(video.videoWidth / (video.videoHeight / 50)); // formula from video server
+        // maximum total possible thumbnails, minus margin to avoid making it feels too narrow
+        const totalThumbnails = Math.round(((this.timeline.current?.offsetWidth ?? 0) - 40 * 2) / thumbnailWidth);
+
+        if (!totalThumbnails) {
+            return (<div className={getClass('timeline-controls')} ref={this.timeline} />);
+        }
+
```

Comment:
Timeline now will be center, so this is to avoid moving left bar when video change (we depend on left bar position in `getPositionInBar` and to calculate which timeline thumbnail to render), kind of place holder